### PR TITLE
Create functional tests

### PR DESCRIPTION
**Pull request by sebastienros**
Added functional tests for
- PostgreSQL
- MongoDB
- MS SQL
- Redis
- RabbitMQ
- MySql

### CODE REVIEW

File: tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using MongoDB.Driver;
+
+public static class MongoDBExtensions
+{
+    public static void ConfigureMongoDB(this WebApplication app)
```
Comment (davidfowl) [inline tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs:8]: Nit: MapMongoMovieApi

**Comment by sebastienros** [comment]
After some feedback I am going to change the tests to use a Fixture representing this AppHost such that each test can use the same app instance and shut it down when the tests are done.

File: tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        await client.ListDatabaseNames().ForEachAsync(databaseNames.Add);
+
+        return databaseNames;
+    }
+
+    private static async Task<List<string>> GetMoviesAsync(IMongoDatabase db)
+    {
+        db.CreateCollection("movies");
+        var moviesCollection = db.GetCollection<Movie>("movies");
+        moviesCollection.InsertOne(new Movie(1, "Rocky I"));
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs:32]: Right, will change that, didn't realize when I moved it from the top level code (so there is nothing running for MongoDB until it's actually invoked).

File: tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+public class RedisFunctionalTests
+{
+    private readonly IntegrationServicesFixture _integrationServicesFixture;
+
+    public RedisFunctionalTests(IntegrationServicesFixture integrationServicesFixture)
+    {
+        _integrationServicesFixture = integrationServicesFixture;
+    }
+
+    [LocalOnlyFact()]
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs:19]: Will look into that, it's possible that it just works, need to check what are the queue requirements on the CI agent.

File: tests/Aspire.Hosting.Tests/TestProgramFixture.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    public override TestProgram CreateTestProgram()
+    {
+        var testProgram = TestProgram.Create<DistributedApplicationTests>(includeNodeApp: true);
+
+        testProgram.AppBuilder.Services
+            .AddHttpClient()
+            .ConfigureHttpClientDefaults(b =>
+            {
+                b.UseSocketsHttpHandler((handler, sp) => handler.PooledConnectionLifetime = TimeSpan.FromSeconds(5));
+                b.AddStandardResilienceHandler();
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/TestProgramFixture.cs:92]: Don't know, it was already there in the initial file. @davidfowl

File: tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs [hunk 5]
```diff
@@ -351,7 +351,7 @@ public async Task VerifyHealthyOnIntegrationServiceA()
 
         await using var app = testProgram.Build();
 
-        var client = app.Services.GetRequiredService<IHttpClientFactory>().CreateClient();
+        using var client = app.Services.GetRequiredService<IHttpClientFactory>().CreateClient();
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:354]: Yes, more things will use fixtures instead. Will do in this PR. Wanted some feedback before I go on.

File: tests/Aspire.Hosting.Tests/Helpers/AllocatedEndpointAnnotationTestExtensions.cs [hunk 6]
```diff
@@ -16,25 +27,70 @@ public static async Task<string> HttpGetAsync<T>(this IResourceBuilder<T> builde
...
+    /// <param name="builder">The resource.</param>
+    /// <param name="client">The <see cref="HttpClient"/> instance to use.</param>
+    /// <param name="bindingName">The name of the binding.</param>
+    /// <param name="path">The path the request is sent to.</param>
+    /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+    /// <returns>The response message.</returns>
+    public static async Task<HttpResponseMessage> HttpGetAsync<T>(this IResourceBuilder<T> builder, HttpClient client, string bindingName, string path, CancellationToken cancellationToken)
+        where T : IResourceWithBindings
+    {
+        // We have to get the allocated endpoint each time through the loop
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/Aspire.Hosting.Tests/Helpers/AllocatedEndpointAnnotationTestExtensions.cs:43]: The initial commit has the comment in a loop:

https://github.com/dotnet/aspire/blob/ef65d63281373fd0322ce683877df22a6d84462b/tests/Aspire.Hosting.Tests/Helpers/AllocatedEndpointAnnotationTestExtensions.cs#L16-L26

Then https://github.com/dotnet/aspire/pull/973 pulled the comment out of the loop into another method.

cc @mitchdenny

File: tests/testproject/TestProject.IntegrationServiceA/Redis/RedisExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        app.MapPost("/redis/{key}", SetKeyAsync);
+
+        app.MapGet("/redis/{key}", GetKeyAsync);
+    }
+
+    private static async Task<IResult> SetKeyAsync(string key, HttpContext context, IConnectionMultiplexer cm)
+    {
+        using var sr = new StreamReader(context.Request.Body);
+        var body = await sr.ReadToEndAsync();
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply tests/testproject/TestProject.IntegrationServiceA/Redis/RedisExtensions.cs:18]: Looks like it only supports JSON:

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/parameter-binding?view=aspnetcore-8.0#binding-failures

Either that, or you can just have a `Stream body` parameter, which is the same as saying `context.Request.Body`.

So I guess this is more complicated than what it is worth. We can just keep the existing code.

File: tests/Aspire.Hosting.Tests/TestProgramFixture.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/TestProgramFixture.cs]: I tried without, then added it back again since some services are not actually ready when the fixture starts. This resilience configuration will have the httpclient retry the requests, which then succeeds the tests.

File: tests/Aspire.Hosting.Tests/Helpers/AllocatedEndpointAnnotationTestExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply tests/Aspire.Hosting.Tests/Helpers/AllocatedEndpointAnnotationTestExtensions.cs:43]: My bad!

File: tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs:19]: There is also a potential issue for port conflicts for some of the integration tests.

File: tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+public static class MongoDBExtensions
+{
+    public static void MapMongoMovieApi(this WebApplication app)
+    {
+        app.MapGet("/mongodb/databases", GetDatabaseNamesAsync);
+
+        app.MapGet("/mongodb/movies", GetMoviesAsync);
+    }
+
+    private static async Task<List<string>> GetDatabaseNamesAsync(IMongoClient client, IMongoDatabase db)
```
Comment (mitchdenny) [inline tests/testproject/TestProject.IntegrationServiceA/MongoDB/MongoDBExtensions.cs:15]: These handlers should probably take the abstract version of the resource as a dependency as a well via keyed DI. That way you can test both flavors.

File: tests/testproject/TestProject.IntegrationServiceA/Redis/RedisExtensions.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (captainsafia) [reply tests/testproject/TestProject.IntegrationServiceA/Redis/RedisExtensions.cs:18]: > Either that, or you can just have a Stream body parameter, which is the same as saying context.Request.Body.

I'd recommend `Stream body` but since this is just test code this is fine

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 9]
```diff
@@ -35,7 +35,7 @@ private TestProgram(string[] args, Assembly assembly, bool includeIntegrationSer
             var sqlserverDbName = "tempdb";
             var mysqlDbName = "mysqldb";
             var postgresDbName = "postgresdb";
-            var mongoDbName = "mongodb";
+            var mongoDbName = "mymongodb";
```
Comment (mitchdenny) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:38]: Why this change?

File: tests/Aspire.Hosting.Tests/MongoDB/MongoDBFunctionalTests.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    public async Task VerifyMongoWorks()
+    {
+        var testProgram = _integrationServicesFixture.TestProgram;
+        var client = _integrationServicesFixture.HttpClient;
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(1));
+
+        var response = await testProgram.IntegrationServiceABuilder!.HttpGetAsync(client, "http", "/mongodb/verify", cts.Token);
+
+        Assert.True(response.IsSuccessStatusCode);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Tests/MongoDB/MongoDBFunctionalTests.cs:29]: When the test fails, what gets logged? It might help to dump the `Results.Problem` information to the test log.

File: tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    public async Task VerifyRedisWorks()
+    {
+        var testProgram = _integrationServicesFixture.TestProgram;
+        var client = _integrationServicesFixture.HttpClient;
+        var data = "Hello World!";
+
+        using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(1));
+
+        await testProgram.IntegrationServiceABuilder!.HttpPostAsync(client, "http", "/redis/hello", new StringContent(data), cts.Token);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs:28]: How come Redis doesn't follow the same pattern as the others with just a `/verify` endpoint?

File: tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs [hunk 12]
```diff
@@ -315,6 +211,8 @@ public async Task TestServicesWithMultipleReplicas()
 
         await app.StartAsync(cts.Token);
 
+        await Task.Delay(1000, cts.Token);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:214]: Why this change? This seems brittle.

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 13]
```diff
@@ -32,12 +32,24 @@ private TestProgram(string[] args, Assembly assembly, bool includeIntegrationSer
...
-            var postgresContainer = AppBuilder.AddPostgresContainer("postgrescontainer");
+            var postgresContainer = AppBuilder.AddPostgresContainer("postgrescontainer")
+                .WithEnvironment("POSTGRES_DB", postgresDbName)
+                .AddDatabase(postgresDbName);
             var rabbitmqContainer = AppBuilder.AddRabbitMQContainer("rabbitmqcontainer");
-            var mongodbContainer = AppBuilder.AddMongoDBContainer("mongodbcontainer");
+            var mongodbContainer = AppBuilder.AddMongoDBContainer("mongodbcontainer")
+                .AddDatabase(mongoDbName);
+
             var sqlserverAbstract = AppBuilder.AddSqlServerContainer("sqlserverabstract");
```
Comment (eerhardt) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:53]: @mitchdenny - what are these "Abstract" containers? It looks to be doing the same thing as above - `AppBuilder.AddSqlServerContainer`, just with a different name.

File: tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs (hunk 12 above)
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:214]: When I looked at the logs (Console) in the test results it was full of retries from the next httpclient calls (retry policy) because the requests are sent before the application is ready. I added this one just to delay when the requests start to limit how many useless queries are sent, and reduce the log size. It doesn't impact the test itself (IMO).

Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:217]: I added a comment

File: tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs (hunk 11 above)
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/Redis/RedisFunctionalTests.cs]: Fixed

File: tests/Aspire.Hosting.Tests/MongoDB/MongoDBFunctionalTests.cs (hunk 10 above)
Comment (sebastienros) [reply tests/Aspire.Hosting.Tests/MongoDB/MongoDBFunctionalTests.cs]: I changed it to render the problem details as the assertion message

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for this great work.

It might be good to get @mitchdenny to sign off on the changes as well.
